File: javascript/ev3.js
var fs = require('fs');

//paths
var ledpath = '/sys/class/leds/ev3-';
var motorPath = '/sys/class/tacho-motor/'; 
var sensorPath = '/sys/class/lego-sensor/';
var soundPath = '/sys/devices/platform/snd-legoev3/';

// changes on board LEDs ex. writeLED('right1','green',20)
function writeLED(location,color,rgb){
    if ((typeof rgb) != 'string') rgb = rgb.toString();
    var path = ledpath + location + ':' + color + ':ev3dev/brightness';
    fs.writeFileSync(path, rgb);
    return;
}

// writes to motor ex. runMotor('A',100)
function runMotor(port,speed){
    port = port.toUpperCase();
    var num = getPortNumber(port);
    if (num === "") return;
    if ((typeof speed) != 'string') speed = speed.toString();
    var MOTORS = getMotors();
    if (port in MOTORS){
        motorSpeedPath = motorPath + MOTORS[port] + '/duty_cycle_sp';
        motorRunPath = motorPath + MOTORS[port] + '/command';
        fs.writeFileSync(motorSpeedPath, speed + '\n');
        fs.writeFileSync(motorRunPath, 'run-forever');
    }
    else{
        console.log("Run Motor -> no motor connected to port " + port)   
    }
    return;
}

// stops motor ex. stopMotor('A')
function stopMotor(port){
    port = port.toUpperCase();
    var num = getPortNumber(port);
        if (num === "") return;
    var MOTORS = getMotors();
    if (port in MOTORS){
        motorRunPath = motorPath + MOTORS[port] + '/command';
        fs.writeFileSync(motorRunPath, 'stop');
    }
    else {
        console.log("Stop Motor -> No motor connected to port " + port);   
    }
    return;
}

// get sensor values ex. getSensor(1)
// returns array of [value, sensor type, units]
function getSensor(port){
    SENSORS = getAllSensors();
    if ((typeof port) != 'string') port = port.toString();
    if (port in SENSORS){
        var sensorValuePath = sensorPath + SENSORS[port] + '/value0';
        var sensorNamePath = sensorPath + SENSORS[port] + '/driver_name';
        try{
            name = fs.readFileSync(sensorNamePath, 'utf8').trim();
            data = sensorInfo(name);
            value = Number(fs.readFileSync(sensorValuePath,'utf8').trim());
            return [value, data[0], data[1]];
           }
        catch(e){console.log('failed to read sensor');}
    }
    return;
}

function playSound(freq,volume){
    var vPath = soundPath + 'volume';
    var freqPath = soundPath + 'tone';
    fs.writeFileSync(vPath, volume.toString() + '\n');
    fs.writeFileSync(freqPath, freq.toString() +'\n');
}

function stopSound(){
    var freqPath = soundPath + 'tone';
    fs.writeFileSync(freqPath, (0).toString() +'\n');
}

// creates motor object
function getMotors(){
    var existingMotors = fs.readdirSync(motorPath);
    var MOTORS = new Object();
    for (var i = 0; i < existingMotors.length; i++){
        var motorPort = motorPath + existingMotors[i] + "/address";
        try{
            var motorRead = fs.readFileSync(motorPort, "utf8");
            MOTORS[motorRead[3]] = existingMotors[i];
        }
        catch(e){}
    }
    return MOTORS;
}

// creates sensor object
function getAllSensors(){
    var SENSORS = new Object();
    // need an error check here if dir doesnt exist! ***********************************
    try {
        var existingSensors = fs.readdirSync(sensorPath);
        for (var i = 0; i < existingSensors.length; i++){
            var sensorPort = sensorPath + existingSensors[i] + '/port_name';
            try{
                var sensorRead = fs.readFileSync(sensorPort,'utf8')
                SENSORS[sensorRead[2]] = existingSensors[i];
            }
            catch(e){console.log("found no sensors")}
        }
    }
    catch(e){
    }
    return SENSORS;
}

function getPortNumber(port){
    var num;
    switch(port){
        case "A": num = 0; break;
        case "B": num = 1; break;
        case "C": num = 2; break;
        case "D": num = 3; break;
        default: num = ""; break;
    }
    return num;
}

// based on returned sensor name, translate name and get units
function sensorInfo(name){
    if ((name).match('lego-ev3-us') || (name).match('lego-nxt-us')) return ['ultrasonic sensor','cm'];            
    // gyro sensor   
    else if ((name).match('lego-ev3-gyro')) return ['gyro sensor','degrees'];                            
    // touch sensor 
    else if ((name).match('lego-ev3-touch') || (name).match('lego-nxt-touch')) return ['touch sensor','boolean'];                           
    // ir sensor 
    else if ((name).match('lego-ev3-ir')) return ['IR sensor','percent'];                                          
    // color sensor
    else if ((name).match('lego-ev3-color')) return ['color sensor','percent'];
    // unknown sensor
    else return ['unknown sensor',''];
}

File: python/ev3.py
# ev3dev python library for LEDs, sensors, and motors
# current functions:
#   writeLED(location,color,rgb)
#   runMotor(port,speed)
#   stopMotor(port)
#   getSensor(port)
#       returns int value, driver name, and mode
#       examples: ultrasonic, gyro, touch, IR, color
#   playSound(freq,volume)
#   stopSound()

import os,sys
from os import curdir,sep,listdir
import re


#LED path
ledpath = '/sys/class/leds/ev3-{}:{}:ev3dev/'
ledbright = ledpath +'brightness'

#motor definitions
motorAttached = '/sys/class/tacho-motor/'
motorpath = '/sys/class/tacho-motor/{}/'
setMotorSpeed = motorpath + 'duty_cycle_sp'
runMotorPath = motorpath + 'command'
checkMotorPort = motorpath + 'address'

#sensor definitions
sensorpath = '/sys/class/lego-sensor/{}/'
sensorValue = sensorpath + 'value0'
sensorAttached = '/sys/class/lego-sensor/'
checkSensorPort = sensorpath + 'address'
sensorModesPath = sensorpath + 'modes'
sensorModePath = sensorpath + 'mode'
drivername = sensorpath + 'driver_name'

soundpath = '/sys/devices/platform/snd-legoev3/{}'


# turn on/off LED ex. writeLED('left1','green',20)
# *********** maybe just left/right and convert? *******
def writeLED(location,color,rgb):
    if ~(type(rgb) is str):
        rgb = str(rgb)
    LED = open(ledbright.format(location,color),"w",0)
    LED.write(rgb + '\n')
    LED.close

def getMotors():
    existingMotors = os.listdir(motorAttached)
    MOTORS = {}
    for i in range(0,len(existingMotors)):
        try:
            motorRead = open(checkMotorPort.format(existingMotors[i]))
            mo = motorRead.read()
            motorRead.close
            MOTORS[mo[3]] = existingMotors[i]
        except IOError:
            print "no motor"
    return MOTORS

def getAllSensors():
    SENSORS = {}
    try:
        existingSensors = os.listdir(sensorAttached)
        for i in range(0,len(existingSensors)):
            try:
                senRead = open(checkSensorPort.format(existingSensors[i]))
                mo = senRead.read()
                senRead.close
                SENSORS[mo[2]] = existingSensors[i]
            except IOError:
                print "no sensors"
    except IOError:
        print "no sensors"
    return SENSORS

def getPortNumber(port):
    if port == "A":
        num = 0
    elif port == "B":
        num = 1
    elif port == "C":
        num = 2
    elif port == "D":
        num = 3
    else:
        num = ''
    return num

def sensorInfo(name):
    if name.find('lego-ev3-us') == 0 or name.find('lego-nxt-us') == 0:
        return 'ultrasonic sensor','cm'
    # gyro sensor
    elif name.find('lego-ev3-gyro') == 0:
        return 'gyro sensor','degrees'
    # touch sensor
    elif name.find('lego-ev3-touch') == 0 or name.find('lego-nxt-touch') == 0:
        return 'touch sensor','boolean'
    # ir sensor
    elif name.find('lego-ev3-ir') == 0:
        return 'IR sensor','percent'
    # color sensor
    elif name.find('lego-ev3-color') == 0:
        return 'color sensor','percent'
    # unknown sensor
    else:
        return 'unknown sensor',''

# turn motor on ex. runMotor('A',100)
def runMotor(port,speed):
    port = port.upper()
    num = getPortNumber(port)
    if num == "":
        print "please enter correct motor port"
        return
    if ~(type(speed) is str):
        speed = str(speed)
    MOTORS = getMotors()
    if port in MOTORS:
        motor = open(setMotorSpeed.format(MOTORS[port]),"w",0)
        motor.write(speed + '\n')
        motor.close
        motor = open(runMotorPath.format(MOTORS[port]),"w",0)
        motor.write('run-forever')
        motor.close
    else:
        print "no motor attached to port {}".format(port)

def stopMotor(port):
    num = getPortNumber(port)
    if num == "":
        print "please enter correct motor port"
        return
    MOTORS = getMotors()
    if port in MOTORS:
        motor = open(runMotorPath.format(MOTORS[port]),"w",0)
        motor.write("stop")
        motor.close
    else:
        print "no motor attached to port {}".format(port)

def getSensor(port):
    SENSORS = getAllSensors()
    if ~(type(port) is str):
        port = str(port)
    if port in SENSORS:
        s = open(sensorValue.format(SENSORS[port]))
        value = s.read()
        s.close
        s = open(drivername.format(SENSORS[port]))
        sensor = s.read();
        s.close
        # name = sensorInfo(sensor)
        units = getSensorMode(port)
        return int(value.strip()),sensor.strip(),units.strip()

def getSensorModes(port):
    SENSORS = getAllSensors()
    if ~(type(port) is str):
        port = str(port)
    if port in SENSORS:
        s = open(sensorModesPath.format(SENSORS[port]))
        value = s.read()
        s.close
        return value

def setSensorMode(port,mode):
    SENSORS = getAllSensors()
    if ~(type(port) is str):
        port = str(port)
    if port in SENSORS:
        s = open(sensorModePath.format(SENSORS[port]),"w",0)
        try:
            s.write(mode + '\n')
        except IOError:
            print IOError
        s.close

def getSensorMode(port):
    SENSORS = getAllSensors()
    if ~(type(port) is str):
        port = str(port)
    if port in SENSORS:
        s = open(sensorModePath.format(SENSORS[port]))
        value = s.read()
        s.close
        return value

def playSound(freq,volume):
    s = open(soundpath.format('volume'),"w",0)
    s.write(str(volume) + '\n')
    s.close()
    s = open(soundpath.format('tone'),"w",0)
    s.write(str(freq) + '\n')

def stopSound():
    s = open(soundpath.format('tone'),"w",0)
    s.write(str(0) + '\n')
